### PR TITLE
[SPARK-50648][CORE] Cleanup zombie tasks in non-running stages when the job is cancelled

### PR DESCRIPTION
**Pull request by yabola**
### What changes were proposed in this pull request?

This is a problem that Spark always had. See the following section for the scenario when the problem occurs.

 When cancel a job, some tasks may be still running. 
The reason is that when `DAGScheduler#handleTaskCompletion` encounters FetchFailed, `markStageAsFinished` will be called to remove the stage in `DAGScheduler#runningStages` (see https://github.com/apache/spark/blob/7cd5c4a1d1eb56fa92c10696bdbd8450d357b128/core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala#L2059) and don't `killAllTaskAttempts`.
But `DAGScheduler#cancelRunningIndependentStages` only find `runningStages`, this will leave zombie shuffle tasks, occupying cluster resources.

### Why are the changes needed?

Assume a job is stage1-> stage2, when FetchFailed occurs during the stage 2, the stage1 and stage2  will resubmit (stage2 may still have some tasks running even if stage2 is resubmitted , this is as expected, because these tasks may eventually succeed and avoid retry)

But during the execution of the stage1-retry , if the SQL is canceled, the tasks in stage1 and stage1-retry can all be killed, but the tasks previously running in stage2 are still running and can't be killed. These tasks can greatly affect cluster stability and occupy resources.

### Does this PR introduce _any_ user-facing change?
No

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

**Comment by wangyum** [comment]
cc @cloud-fan

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala [hunk 1]
```diff
@@ -2937,7 +2937,9 @@ private[spark] class DAGScheduler(
         } else {
           // This stage is only used by the job, so finish the stage if it is running.
           val stage = stageIdToStage(stageId)
-          if (runningStages.contains(stage)) {
+          val isRunningStage = runningStages.contains(stage) ||
+            (waitingStages.contains(stage) && taskScheduler.hasRunningTasks(stageId))
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:2941]: what if we just kill all waiting stages? Does `taskScheduler.killAllTaskAttempts` handle it well?

**Comment by cloud-fan** [comment]
This is a good catch! cc @jiangxb1987 @Ngone51

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala (hunk 1 above)
Comment (cloud-fan) [reply core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:2941]: and can we add a special flag to indicate the waiting stages that are submitted due to retry?

Comment (yabola) [reply core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala]: > what if we just kill all waiting stages? Does taskScheduler.killAllTaskAttempts handle it well?
I tested that: if the normally generated waiting stages call `killAllTaskAttempts`,  the stage status will be displayed as `FAILED`, which was `SKIPPED` before, killAllTaskAttempts  itself will not go wrong.
Always kill waiting stage seems to be a safer approach (tasks shouldn't be run anymore), but it may generate unnecessary `stageFailed` events compared with before.

> and can we add a special flag to indicate the waiting stages that are submitted due to retry?
Yes, we can add a flag , please see the update codes. 
Actually , there is a trade-off here to kill waiting stages
- always kill
- kill stages had failed ( `stage#failedAttemptIds` > 0)
- kill stages failed when fetch failed (`stage#resubmitInFetchFailed`)
- kill stages only have running tasks

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala [hunk 2]
```diff
@@ -2248,7 +2250,7 @@ class DAGSchedulerSuite extends SparkFunSuite with TempLocalSparkContext with Ti
     // original result task 1.0 succeed
     runEvent(makeCompletionEvent(taskSets(1).tasks(1), Success, 42))
     sc.listenerBus.waitUntilEmpty()
-    assert(completedStage === List(0, 1, 1, 0))
+    assert(completedStage === List(0, 1, 1, 0, 1))
```
Comment (yabola) [inline core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala:2253]: The last one added is the resubmit stage (FetchFailed) and in waiting stages. We will kill it and one more SparkListenerStageCompleted event will be added ( see markStageAsFinished)

File: core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala [hunk 3]
```diff
@@ -1212,6 +1212,16 @@ private[spark] class TaskSchedulerImpl(
...
   override def applicationAttemptId(): Option[String] = backend.applicationAttemptId()
 
+  override def hasRunningTasks(stageId: Int): Boolean = synchronized {
+    var hasRunningTasks = false
+    taskSetsByStageIdAndAttempt.get(stageId).foreach { attempts =>
+      attempts.foreach { case (_, tsm) =>
+        hasRunningTasks = hasRunningTasks || tsm.runningTasksSet.nonEmpty
+      }
+    }
+    hasRunningTasks
```
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala:1222]:
```suggestion
    taskSetsByStageIdAndAttempt.get(stageId).exists{ attempts =>
      attempts.exists(_._2.runningTasksSet.nonEmpty)
    }
```

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala [hunk 4]
```diff
@@ -2187,7 +2187,8 @@ private[spark] class DAGScheduler(
                   log"we will roll back and rerun below stages which include itself and all its " +
                   log"indeterminate child stages: ${MDC(STAGES, rollingBackStages)}")
               }
-
+              failedStage.markResubmitInFetchFailed()
+              mapStage.markResubmitInFetchFailed()
```
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:2191]: nit: `markResubmitInFetchFailed` -> `setResubmitByFetchFailure`.

Also, please reset this to `false` - when stage transitions to running - so that scheduler state is consistent.
`submitMissingTasks` preferably, or in `submitStage`

File: core/src/main/scala/org/apache/spark/scheduler/TaskScheduler.scala [hunk 5]
```diff
@@ -121,4 +121,6 @@ private[spark] trait TaskScheduler {
    */
   def applicationAttemptId(): Option[String]
 
+
+  def hasRunningTasks(stageId: Int): Boolean
```
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/scheduler/TaskScheduler.scala:125]: Why do we need this method ?

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala (hunk 4 above)
Comment (yabola) [reply core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala]: Done,  this can make the variable meaning more accurate. 
But in addition, when `submittingMissingTasks`,  the stage will be added to the `runningStages`, we can always cancel job normally.

File: core/src/main/scala/org/apache/spark/scheduler/TaskScheduler.scala (hunk 5 above)
Comment (yabola) [reply core/src/main/scala/org/apache/spark/scheduler/TaskScheduler.scala:125]: [Please see](https://github.com/apache/spark/pull/49270#discussion_r1896703317)
This is the code I used to complete this in the beginning (kill waiting stage which only has running tasks). I'm not sure which way is better.

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala [hunk 6]
```diff
@@ -2937,7 +2938,9 @@ private[spark] class DAGScheduler(
         } else {
           // This stage is only used by the job, so finish the stage if it is running.
           val stage = stageIdToStage(stageId)
-          if (runningStages.contains(stage)) {
+          val shouldKill = runningStages.contains(stage) ||
+            (waitingStages.contains(stage) && stage.resubmitInFetchFailed)
```
[the remark this answers is not in the document]
Comment (yabola) [reply core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:2943]: Yes we can [Please see this](https://github.com/apache/spark/pull/49270#discussion_r1896703317)
 I'm not sure which way is better.

File: core/src/main/scala/org/apache/spark/scheduler/TaskScheduler.scala (hunk 5 above)
Comment (mridulm) [reply core/src/main/scala/org/apache/spark/scheduler/TaskScheduler.scala:125]: If the method is no longer being used, please remove it.

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (mridulm) [reply core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:2943]: I like @Ngone51's suggestion better - simply check for `stage.failedAttemptIds.nonEmpty || runningStages.contains(stage)`

[the remark this answers is not in the document]
Comment (yabola) [reply core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:2943]: @mridulm @Ngone51 do you think it is necessary `(waitingStages.contains(stage) && stage.failedAttemptIds.nonEmpty) || runningStages.contains(stage)`. Only considering `failedAttemptIds` may result in repeated calls to the the stage already completed and failed.

[the remark this answers is not in the document]
Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:2943]: > Only considering failedAttemptIds may result in repeated calls to the the stage already completed and failed.

It looks like there could be a case where the stage exists in `failedStages` but not in `waitingStages`, e.g., in the case of fetch failures, map stage and reduce stage can be added into `failedStages`, but the related job could be canceled before they were resubmitted. So adding `waitingStages.contains(stage)` would miss the stages in `failedStages`. And I don't think we would have repeated calls as we don't kill tasks for those failed stages.

[the remark this answers is not in the document]
Comment (yabola) [reply core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala]: Thanks for the confirmation, done

File: sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala [hunk 7]
```diff
@@ -938,8 +938,12 @@ class AdaptiveQueryExecSuite
...
             joined.collect()
           }
-          assert((Seq(error) ++ Option(error.getCause) ++ error.getSuppressed()).exists(
-            e => e.getMessage() != null && e.getMessage().contains("coalesce test error")))
+          val errorMessages = (Seq(error) ++ Option(error.getCause) ++ error.getSuppressed())
+            .filter(e => e.getMessage != null).map(e => e.getMessage)
+          assert(errorMessages.exists(
+            e => e.contains("coalesce test error")),
+            s"Error messages should contain `coalesce test error`, " +
+              s"error messages: $errorMessages")
```
Comment (Ngone51) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala:946]: Unrelated changes?

Comment (yabola) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala:946]: It is unrelated, but this seems to be an unstable UT. I want to print clear information when an error occurs.

Comment (Ngone51) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala:946]: You can fix it separately. But please exclude the unrelated changes in this PR.

Comment (yabola) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala]: ok ,I removed

**Review by Ngone51 — APPROVED** [review]

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala (hunk 2 above)
Comment (yabola) [reply core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala]: After changing to judging by `failedAttemptIds`,  it won't  cancel. Because all tasks finished already , and in `markStageAsFinished` will remove `failedAttemptIds` if no error message.

Comment (Ngone51) [reply core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala]: @yabola I'm curious about the difference here. With the current approach, doesn't the stage still has to be killed because of `failedAttemptIds.nonEmpty`?

Comment (yabola) [reply core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala]: In addition, the result stage will definitely kill all tasks after success, please see here
https://github.com/apache/spark/blob/939129ec01af7f7b6dbec737f2d4149d2fc0d9a3/core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala#L1960

Comment (Ngone51) [reply core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala]: OK, sounds good to me.

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala [hunk 8]
```diff
@@ -206,12 +208,16 @@ class DAGSchedulerSuite extends SparkFunSuite with TempLocalSparkContext with Ti
       // normally done by TaskSetManager
       taskSet.tasks.foreach(_.epoch = mapOutputTracker.getEpoch)
       taskSets += taskSet
+      val taskPartitionIds = new HashSet[Int]()
+      taskSet.tasks.foreach(task => taskPartitionIds += task.partitionId)
+      runningTaskInfos.put(taskSet.stageId, taskPartitionIds)
```
Comment (mridulm) [inline core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala:213]: nit:
```suggestion
      runningTaskInfos.put(taskSet.stageId, new HashSet[Int]() ++ taskSet.tasks..map(_.partitionId))
```

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala [hunk 9]
```diff
@@ -393,6 +399,16 @@ class DAGSchedulerSuite extends SparkFunSuite with TempLocalSparkContext with Ti
...
     }
+
+    override private[scheduler] def handleTaskCompletion(event: CompletionEvent): Unit = {
+      super.handleTaskCompletion(event)
+      if (runningTaskInfos.contains(event.task.stageId)) {
+        runningTaskInfos(event.task.stageId) -= event.task.partitionId
+        if (runningTaskInfos(event.task.stageId).isEmpty) {
+          runningTaskInfos.remove(event.task.stageId)
+        }
+      }
```
Comment (mridulm) [inline core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala:410]: nit: 
```suggestion
      runningTaskInfos.get(event.task.stageId).foreach{ partitions =>
        partitions -= event.task.partitionId
        if (partitions.isEmpty) runningTaskInfos.remove(event.task.stageId)
      }
```

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala [hunk 10]
```diff
@@ -2252,6 +2268,46 @@ class DAGSchedulerSuite extends SparkFunSuite with TempLocalSparkContext with Ti
     assert(scheduler.activeJobs.isEmpty)
   }
 
+  test("SPARK-50648: when job is cancelled during shuffle retry in parent stage, " +
```
Comment (mridulm) [inline core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala:2271]: Nice test !

**Review by mridulm — COMMENTED** [review]
Looks good to me, just a few nits.

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala (hunk 10 above)
Comment (yabola) [reply core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala:2267]: Thanks, done

**Comment by Ngone51** [comment]
Thanks, merged to master!

**Comment by yabola** [comment]
Thank you all for the review
